### DISCUSSION TITLE
Make `io` module a builtin in Nelua

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#123] [Ideas]
With Lua I use either `string.format()` to format my output or go with `io.write()`.

When I want to format output in Nelua, I'm forced to `require` either `string` or `io`.

Can we make at least `io` part of Nelua's core? It could help a lot.

Cheers.

**Comment by edubart** [comment]
Actually I'm currently studying a way to do this, because I want to move `print`, `assert`, `error`, and other builtin functions to pure Nelua code, and making them use `io` is an option, but there are few problems at the moment that I am looking to solve:

1. Not all applications uses `io`, for example in AVR using `io` does not make sense, also would break compilation because the `io` module always imports symbols not available in AVR like `stdout`.
2. Always importing `io` would always import the `GC`, but I've made some workarounds that, I am still experimenting though.
3. Always importing `io` would impact compile speed for applications that don't use `io` module, are you fine with this?

I will keep researching a way to do this, with all that in mind, thanks for the suggestion.

**Reply by stefanos82** [reply]
> Actually I'm currently studying a way to do this, because I want to move `print`, `assert`, `error`, 
> and other builtin functions to pure Nelua code, and making them use `io` is an option, but there are
> few problems at the moment that I am looking to solve:
> 
> 1. Not all applications uses `io`, for example in AVR using `io` does not make sense, also would break compilation because the `io` module always imports symbols not available in AVR like `stdout`.

So AVR microcontrollers don't have access to `I/O`?

> 
> 2. Always importing `io` would always import the `GC`, but I've made some workarounds that, I am still experimenting though.
> 

I have found `uLua` which is suited for AVR 8-bit and seems like they use `io` and `string.format()` [ChunkSpy.lua#L389](https://github.com/deemess/uLua/blob/master/test/ChunkSpy.lua#L389)

Maybe you could advice its C code and see their workarounds?

> 3. Always importing `io` would impact compile speed for applications that don't use `io` module, are you fine with this?

That is something I didn't think it would affect the compilation procedure, as I'm constantly using it in Lua and never noticed the slightest thing.

> I will keep researching a way to do this, with all that in mind, thanks for the suggestion.

No worries.

**Comment by edubart** [comment]
> 1. Not all applications uses `io`, for example in AVR using `io` does not make sense, also would break compilation because the `io` module always imports symbols not available in AVR like `stdout`.
> 2. Always importing `io` would always import the `GC`, but I've made some workarounds that, I am still experimenting though.
> 3. Always importing `io` would impact compile speed for applications that don't use `io` module, are you fine with this?

I've made progress in 1. and 2., now when you do `require 'io'` the C code gen is much smaller, this makes things more approachable to make `io` a core module in the future once I finish some other details.

**Reply by stefanos82** [reply]
Very nice, well done :+1:
